components/concept/storage/src/main/java/mozilla/components/concept/storage/BookmarksStorage.kt

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.storage
+
+interface BookmarksStorage {
+
+    /**
+     * Produces a bookmarks tree for the given guid string.
+     *
+     * @param guid The bookmark guid to obtain.
+     * @param recursive Whether to recurse and obtain all levels of children.
+     * @return The populated root starting from the guid.
+     */
+    suspend fun getTree(guid: String, recursive: Boolean = false): BookmarkNode?
+
+    /**
+     * Searches bookmarks with a query string.
+     *
+     * @param query The query string to search.
+     * @param limit The maximum number of entries to return.
+     * @return The list of matching bookmark nodes up to the limit number of items.
+     */
+    suspend fun searchBookmarks(query: String, limit: Int = defaultBookmarkSearchLimit): List<BookmarkNode>
+
+    /**
+     * Adds a new bookmark item to a given node.
+     *
+     * Sync behavior: will add new bookmark item to remote devices.
+     *
+     * @param parentGuid The parent guid of the new node.
+     * @param newNode The details of the node to add.
+     * @param position The optional position to add the new node or null to append.
+     * @return The guid of the newly inserted bookmark item.
+     */
+    suspend fun addItem(parentGuid: String, newNode: BookmarkNode, position: Int? = null): String
```

Comment:
Same comment here as for folder creation. A bookmark is best described by a parentGuid, url, title, and an optional position.

Reply:
Fixed